### PR TITLE
Don't aggregate rooms and users in SpaceStore

### PR DESCRIPTION
**Pull request by kerryarchibald**
Changes the way SpaceStore stores space data. 
Before: rooms and users aggregated up the space hierarchy, lots of traversal on changes to update state up and down hierarchy
After: rooms and users are stored in a single layer maps with their space key, along with a map of SpaceKeys to contained sub spaces.

Enables hierarchical `SpaceFilterCondition` and then configurable aggregation of subspace rooms and users in the room list.

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Preview: https://pr7723--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/stores/spaces/flattenSpaceHierarchy.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        descendentSpaces?.forEach(
+            descendantSpaceId => {
+                if (!flatSpace.has(descendantSpaceId)) {
+                    traverseSpaceDescendants(spaceDescendantMap)(descendantSpaceId, flatSpace);
+                }
+            },
+        );
+
+        return flatSpace;
+    };
```
Comment (t3chguy) [inline src/stores/spaces/flattenSpaceHierarchy.ts:35]: Any reason for this not to just be a simple method taking 3 params? Its returned function always gets called immediately. 

e.g `traverseSpaceDescendants(spaceDescendantMap, descendantSpaceId, flatSpace);`

File: src/stores/spaces/SpaceStore.ts [hunk 2]
```diff
@@ -99,9 +100,11 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
     private parentMap = new EnhancedMap<string, Set<string>>();
     // Map from SpaceKey to SpaceNotificationState instance representing that space
     private notificationStateMap = new Map<SpaceKey, SpaceNotificationState>();
-    // Map from space key to Set of room IDs that should be shown as part of that space's filter
+    // Map from SpaceKey to Set of room IDs that are direct descendants of that space
     private spaceFilteredRooms = new Map<SpaceKey, Set<string>>(); // won't contain MetaSpace.People
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:104]: think this identifier could do with an update given its new usage

File: src/stores/spaces/SpaceStore.ts [hunk 3]
```diff
@@ -99,9 +100,11 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
-    // Map from space ID to Set of user IDs that should be shown as part of that space's filter
+    // Map from SpaceKey to Set of space keys that are direct descendants of that space
+    private childSpacesBySpace = new Map<SpaceKey, Set<SpaceKey>>(); // won't contain MetaSpace.People
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:106]: Well surely it won't contain any metaspaces, not just People, given metaspaces can't contain other spaces.

File: src/stores/spaces/SpaceStore.ts [hunk 4]
```diff
@@ -350,12 +353,14 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
     }
 
-    public isRoomInSpace(space: SpaceKey, roomId: string): boolean {
+    public isRoomInSpace(space: SpaceKey, roomId: string, includeDescendantSpaces = true): boolean {
         if (space === MetaSpace.Home && this.allRoomsInHome) {
             return true;
         }
 
-        if (this.spaceFilteredRooms.get(space)?.has(roomId)) {
+        console.log(space, this.getSpaceFilteredRoomIds(space, includeDescendantSpaces));
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:361]: stray debugging console log

File: src/stores/spaces/SpaceStore.ts [hunk 5]
```diff
@@ -380,19 +385,31 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
-        if (isMetaSpace(space)) return undefined;
-        return this.spaceFilteredUsers.get(space) || new Set();
+        if (isMetaSpace(space)) {
+            return undefined;
+        }
+        if (!includeDescendantSpaces) {
+            return this.spaceFilteredUsers.get(space) || new Set();
+        }
+
+        return flattenSpaceHierarchy(this.spaceFilteredUsers, this.childSpacesBySpace)(space);
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:412]: The param name for `flattenSpaceHierarchy` is `spaceRoomMap` with type `SpaceRoomMap` yet we're giving it a Map from SpaceKey to Set of userIds, not roomIds.

File: src/stores/spaces/flattenSpaceHierarchy.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    (spaceId: SpaceKey): Set<string> => {
+        const flattenedSpaceIds = traverseSpaceDescendants(spaceDescendantMap)(spaceId);
+        const flattenedRooms = new Set<string>();
+
+        flattenedSpaceIds.forEach(id => {
+            const roomIds = spaceRoomMap.get(id);
+            roomIds?.forEach(flattenedRooms.add, flattenedRooms);
+        });
+
+        return flattenedRooms;
```
Comment (t3chguy) [inline src/stores/spaces/flattenSpaceHierarchy.ts:53]: This kind of looks equivalent to

```suggestion
		return new Set(flattenedSpaceIds.flatMap(id => spaceRoomMap.get(id) ?? []));
```

or similar

File: src/stores/spaces/SpaceStore.ts [hunk 7]
```diff
@@ -380,19 +385,31 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
         return false;
     }
 
-    public getSpaceFilteredRoomIds = (space: SpaceKey): Set<string> => {
+    // get all rooms in a space
+    // including descendant spaces
+    public getSpaceFilteredRoomIds = (space: SpaceKey, includeDescendantSpaces = true): Set<string> => {
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:390]: there are a lot of call sites with `includeDescendantSpaces = true` and no caching, it feels like that may have detrimental performance effects for users with many deep space graphs.

File: src/stores/spaces/flattenSpaceHierarchy.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    };
+
+/**
+ * Helper function to traverse space heirachy and flatten
+ * @param spaceRoomMap ie map of rooms or dms
+ * @param spaceDescendantMap map of spaces and their children
+ * @returns set of all rooms
+ */
+export const flattenSpaceHierarchy = (spaceRoomMap: SpaceRoomMap, spaceDescendantMap: SpaceDescendantMap) =>
+    (spaceId: SpaceKey): Set<string> => {
```
Comment (t3chguy) [inline src/stores/spaces/flattenSpaceHierarchy.ts:44]: Any reason for this to not just be a simple method taking 3 params too? All call sites seem to call immediately and not keep ahold of the returned anonymous function

File: src/stores/spaces/SpaceStore.ts [hunk 9]
```diff
@@ -719,31 +686,40 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
             ...roomDiff.added,
             ...userDiff.added,
+            ...spaceDiff.added,
             ...roomDiff.removed,
             ...userDiff.removed,
+            ...spaceDiff.removed,
             ...roomsChanged,
             ...usersChanged,
+            ...spacesChanged,
         ]);
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:723]: If a subspace's direct descendent room list changes this seems like it'll no longer emit an update on its parent spaces (recursing) which is a behavioural change

File: src/stores/spaces/SpaceStore.ts (hunk 7 above)
Comment (kerryarchibald) [reply src/stores/spaces/SpaceStore.ts:389]: Yeah, I will do some profiling on this. Implementing hierarchical SpaceFilterConditions should remove quite a few of the calls, but there are still quite a few coming from notification state management too.

Comment (t3chguy) [reply src/stores/spaces/SpaceStore.ts:389]: Notification states could be hierarchical also

File: src/stores/spaces/SpaceStore.ts [hunk 10]
```diff
@@ -553,6 +567,9 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
             this.roomIdsBySpace.set(MetaSpace.Orphans, new Set(orphans.map(r => r.roomId)));
         }
 
+        // bust cache for meta spaces
+        enabledMetaSpaces.forEach(metaSpaceKey => this._aggregatedSpaceCache.roomIdsBySpace.delete(metaSpaceKey));
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:571]: Why are we even bothering to aggregate-cache the metaspaces? They always contain no nesting so don't need duplicating in a cache

File: src/stores/spaces/SpaceStore.ts [hunk 11]
```diff
@@ -107,6 +107,12 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
     private childSpacesBySpace: SpaceDescendantMap = new Map<SpaceKey, Set<SpaceKey>>();
     // Map from SpaceKey to Set of user IDs that are direct descendants of that space
     private userIdsBySpace: SpaceEntityMap = new Map<Room["roomId"], Set<string>>();
+    // cache that stores the aggregated lists of roomIdsBySpace and userIdsBySpace
+    // cleared on changes
+    private _aggregatedSpaceCache: Record<string, SpaceEntityMap> = {
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:112]: this type is weird, given the keys are not plain strings, they're just a couple of specific explicit fields, imo would be better without an explicit type defn

File: src/stores/spaces/SpaceStore.ts (hunk 9 above)
Comment (kerryarchibald) [reply src/stores/spaces/SpaceStore.ts:764]: Fixed this using the `parentMap`, here and in `onMembersUpdate`.

Thanks for all your reviews here!

File: src/stores/spaces/SpaceStore.ts (hunk 7 above)
Comment (t3chguy) [reply src/stores/spaces/SpaceStore.ts]: Out of interest was there a reason to not do this or is this a possible future improvement?

Comment (kerryarchibald) [reply src/stores/spaces/SpaceStore.ts]: I have another PR lined up for hierarchical SpaceFilterConditions after this is merged, wanted to limit the scope here. Hierarchical notifications don't directly contribute to my ticket - aggregate room list toggling - so will only be done time permitting.

File: src/stores/spaces/SpaceStore.ts [hunk 12]
```diff
@@ -101,10 +107,18 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
     // Map from SpaceKey to SpaceNotificationState instance representing that space
     private notificationStateMap = new Map<SpaceKey, SpaceNotificationState>();
-    // Map from space key to Set of room IDs that should be shown as part of that space's filter
-    private spaceFilteredRooms = new Map<SpaceKey, Set<string>>(); // won't contain MetaSpace.People
-    // Map from space ID to Set of user IDs that should be shown as part of that space's filter
-    private spaceFilteredUsers = new Map<Room["roomId"], Set<string>>();
+    // Map from SpaceKey to Set of room IDs that are direct descendants of that space
+    private roomIdsBySpace: SpaceEntityMap = new Map<SpaceKey, Set<string>>(); // won't contain MetaSpace.People
+    // Map from SpaceKey to Set of space keys that are direct descendants of that space
+    private childSpacesBySpace: SpaceDescendantMap = new Map<SpaceKey, Set<SpaceKey>>();
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:113]: Using SpaceKey here creates ambiguity, implies that metaspaces can have descendants and that metaspaces can be descendants when that isn't true.

File: src/stores/spaces/SpaceStore.ts [hunk 13]
```diff
@@ -101,10 +107,18 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
+    // Map from SpaceKey to Set of user IDs that are direct descendants of that space
+    private userIdsBySpace: SpaceEntityMap = new Map<Room["roomId"], Set<string>>();
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:115]: userIds aren't tracked by-metaspace, only for real spaces, comment does not match type `Room["roomId"]`

**Review by t3chguy — APPROVED** [review]
Looks sane to me otherwise, thanks!!

**Pull request merged**